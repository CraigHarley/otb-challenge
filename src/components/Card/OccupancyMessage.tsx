import React from 'react';

export interface IOccupancyMessageProps {
    adults: number;
    children: number;
    infants: number;
}

export default ({adults, children, infants}: IOccupancyMessageProps) => {
    if (adults && !children && !infants) {
        return (
            <span>
                {adultMessage(adults)}
            </span>
        );
    }

    if (adults && children && infants) {
        return (
            <span>
                {adultMessage(adults)}, {childMessage(children)} & {infantMessage(infants)}
            </span>
        );
    }

    if (adults && children && !infants) {
        return (
            <span>
                {adultMessage(adults)} & {childMessage(infants)}
            </span>
        );
    }

    if (adults && !children && infants) {
        return (
            <span>
                {adultMessage(adults)} & {infantMessage(infants)}
            </span>
        );
    }

    // todo make this impossible with types.
    throw Error("Can't have a holiday with no adults.");
}

const adultMessage = (count: number) => (
    <>
        <strong>{count}</strong> {pluralize('adult', count)}
    </>
);
const childMessage = (count: number) => (
    <>
        <strong>{count}</strong> {pluralizeChildren(count)}
    </>
);
const infantMessage = (count: number) => (
    <>
        <strong>{count}</strong> {pluralize('infant', count)}
    </>
);

// Could also have used https://www.npmjs.com/package/pluralize
// But it's two lines vs a whole dep.
const pluralize = (original: string, count: number): string => count === 1 ? original : original + 's';
const pluralizeChildren = (count: number): string => count === 1 ? 'child' : 'children';
