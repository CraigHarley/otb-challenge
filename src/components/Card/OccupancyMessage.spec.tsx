import React from 'react';
import {render} from 'enzyme';

import OccupancyMessage, {IOccupancyMessageProps} from './OccupancyMessage';

describe('OccupancyMessage', () => {
    test('should function correctly for all the valid occupancy types',
        () => {
            const occupancies: IOccupancyMessageProps[] = [
                {
                    adults: 2,
                    children: 2,
                    infants: 2
                },
                {
                    adults: 2,
                    children: 2,
                    infants: 1
                },
                {
                    adults: 1,
                    children: 1,
                    infants: 1
                },
                {
                    adults: 2,
                    children: 1,
                    infants: 0
                },
                {
                    adults: 2,
                    children: 1,
                    infants: 0
                },
                {
                    adults: 2,
                    children: 0,
                    infants: 0
                },
            ];

            occupancies
                .forEach((rule, key) =>
                    expect(
                        render(
                            <OccupancyMessage
                                {...rule}
                            />
                        )
                    ).toMatchSnapshot(`Occupancy rule: ${key}`)
                );
        }
    );

    it('should throw an exception if there are no adults',
        () =>
            expect(
                () =>
                    render(
                        <OccupancyMessage
                            adults={0}
                            children={2}
                            infants={0}
                        />
                    )
            )
                .toThrowError('Can\'t have a holiday with no adults.')
    )
})
