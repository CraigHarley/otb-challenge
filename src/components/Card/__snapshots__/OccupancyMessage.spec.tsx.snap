// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OccupancyMessage should function correctly for all the valid occupancy types: Occupancy rule: 0 1`] = `
<span>
  <strong>
    2
  </strong>
   adults, 
  <strong>
    2
  </strong>
   children & 
  <strong>
    2
  </strong>
   infants
</span>
`;

exports[`OccupancyMessage should function correctly for all the valid occupancy types: Occupancy rule: 1 1`] = `
<span>
  <strong>
    2
  </strong>
   adults, 
  <strong>
    2
  </strong>
   children & 
  <strong>
    1
  </strong>
   infant
</span>
`;

exports[`OccupancyMessage should function correctly for all the valid occupancy types: Occupancy rule: 2 1`] = `
<span>
  <strong>
    1
  </strong>
   adult, 
  <strong>
    1
  </strong>
   child & 
  <strong>
    1
  </strong>
   infant
</span>
`;

exports[`OccupancyMessage should function correctly for all the valid occupancy types: Occupancy rule: 3 1`] = `
<span>
  <strong>
    2
  </strong>
   adults & 
  <strong>
    0
  </strong>
   children
</span>
`;

exports[`OccupancyMessage should function correctly for all the valid occupancy types: Occupancy rule: 4 1`] = `
<span>
  <strong>
    2
  </strong>
   adults & 
  <strong>
    0
  </strong>
   children
</span>
`;

exports[`OccupancyMessage should function correctly for all the valid occupancy types: Occupancy rule: 5 1`] = `
<span>
  <strong>
    2
  </strong>
   adults
</span>
`;
